File: 507_Perfect_Number.java
/*
https://leetcode.com/problems/perfect-number/
We define the Perfect Number is a positive integer that is equal to the sum of all its positive divisors except itself.
Now, given an integer n, write a function that returns true when it is a perfect number and false when it is not.

Try all divisors from 2 to sqrt(n).
Time - O(sqrt(n))
Space - O(1)
*/

class Solution {
    public boolean checkPerfectNumber(int num) {

        if (num == 1)       // no divisors apart from self
            return false;

        int sumDivisors = 1;
        int maxDivisor = (int) Math.sqrt(num);
        for (int i = 2; i <= maxDivisor; ++i) {
            if (num % i == 0)
                sumDivisors += i + num/i;
        }

        if (maxDivisor * maxDivisor == num)     // double-counted if num is a square
            sumDivisors -= maxDivisor;

        return sumDivisors == num;
    }
}

File: 507_Perfect_Number.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/perfect-number/
# We define a Perfect Number as a positive integer that is equal to the sum of all its positive divisors except itself.
# Now, given an integer n, write a function that returns true when it is a perfect number and false when it is not.

# Check all divisor integers up to and including sqrt(num). If divisor has no remainder, add divisor and result of
# division.
# Time - O(n**0.5)
# Space - O(1)

import math

class Solution(object):
    def checkPerfectNumber(self, num):
        """
        :type num: int
        :rtype: bool
        """
        if num <= 1:                                    # negative, 0 and 1 are not perfect
            return False

        sum_divisors = 1

        for i in range(2, int(math.sqrt(num)) + 1):     # sqrt slightly faster than num**0.5

            div, mod = divmod(num, i)
            if mod == 0:
                sum_divisors += i + div

        return sum_divisors == num